### DISCUSSION TITLE
Map animations

### DISCUSSION THREAD

**Discussion by Psijic** [mapsforge/vtm#1070] [Q&A] [answered]
Didn't find about animations in samples or manuals (where are they?). How to do the next?

1. Animated zoom. I tried `map.mapPosition.setZoom()` - no result. Use `map.setMapPosition()` but it's not animated.
2. Animated map [camera] position moving. Tried `map.viewport().moveMap()` - no result. Use not animated `map.setMapPosition() `.
3. Get if map [camera] position is moving at the moment (e.g. user drags the map viewport). `cameraPositionState.isMoving` in G-Maps

**Comment by devemux86** [comment, marked as answer]
1,2: `map.animator().`
3: `map.events.bind(UpdateListener)` like in `PathOverlayActivity`, probably listen for `Map.MOVE_EVENT`

**Reply by Psijic** [reply]
I need to save map position after it was moved. Could you help with the right event combination shows map position stopped being changed?  
```
  map.events.bind(Map.UpdateListener { e, mapPosition ->
        if (e == Map.MOVE_EVENT || e == Map.ANIM_END)
            viewModel.setCameraPosition(mapPosition.latitude, mapPosition.longitude, mapPosition.zoom)
    })
```
I hope, there is no need to use double check like first on the action start and second on the action end. This is how I did it with GM Compose:
```
LaunchedEffect(cameraPositionState.isMoving) {
    if (!cameraPositionState.isMoving) 
   ...
}
```

**Reply by Psijic** [reply]
Also not sure why `animateZoom` uses `scaleBy` only but not the final zoom level as an option.

**Reply by devemux86** [reply]
Check for `if (e == Map.UPDATE_EVENT)`, like the [MapScaleBarLayer](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/scalebar/MapScaleBarLayer.java).

**Reply by Psijic** [reply]
Nope. False positive.

**Reply by Psijic** [reply]
Something like this is needed ATM:

```
@Composable
private fun MapMovingProceed(map: Map) {
    var isMoving by remember { mutableStateOf(false) }

    LaunchedEffect(isMoving) {
        if (isMoving) {
            Timber.d("my code here")
        }
    }

    DisposableEffect(map) {
        val listener = Map.UpdateListener { e, mapPosition ->
            when (e) {
                Map.MOVE_EVENT, Map.ANIM_START -> {
                    // Set isMoving to true to indicate that the map is currently moving
                    isMoving = true
                }

                Map.ANIM_END -> {
                    // Check if the map was previously moving (isMoving is true)
                    if (isMoving) {
                        // Reset isMoving to false
                        isMoving = false
                    }
                }
            }
        }

        // Bind the event listener when the Composable is first composed
        map.events.bind(listener)

        // Unbind the event listener when the Composable is no longer in use
        onDispose {
            map.events.unbind(listener)
        }
    }
}
```
